crates/wit-component/src/linking/metadata.rs

```diff
@@ -344,8 +345,16 @@ impl<'a> Metadata<'a> {
                 Payload::TypeSection(reader) => {
                     types = reader
                         .into_iter()
-                        .filter_map(|r| {
-                            r.map(|ty| match ty.structural_type {
+                        .flat_map(|r| match r {
+                            Err(e) => vec![Err(e)],
+                            Ok(rg) => rg
+                                .types()
+                                .iter()
+                                .map(|st| Ok(st.clone()))
+                                .collect::<Vec<_>>(),
+                        })
```

Comment:
We shouldn't force additional `Vec` allocations here. It is fine to write a `for` loop if the iterators are pushing you in this direction.

Reply:
Please don't kill it! I spent entire 3 days trying to come up with it 😅

Interesting though, the current code in this PR is slightly faster than this:
```rust
                    types = Vec::new();
                    for rg in reader {
                        for ty in rg?.types() {
                            if let StructuralType::Func(ty) = ty.clone().structural_type {
                                types.push(ty);
                            }
                        }
                    }
```

If we use this snippet, performance drops:
```
parse/spidermonkey      time:   [15.602 ms 15.622 ms 15.641 ms]                               
                        change: [+1.5950% +1.8649% +2.1264%] (p = 0.00 < 0.05)
                        Performance has regressed.

validate/lots-of-types  time:   [741.42 µs 744.22 µs 747.13 µs]                                   
                        change: [+5.1396% +5.5738% +5.9996%] (p = 0.00 < 0.05)
                        Performance has regressed.

parse/lots-of-types     time:   [164.25 µs 164.57 µs 164.87 µs]                                
                        change: [+1.8325% +2.3509% +2.8039%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

Reply:
Huh! Okay well I can't argue with profiling results!